File: src/ffprobe.py
from importlib.resources import path
import json
from . import path
import shlex
from . import proc


def stream_info_json(input_file_path: str) -> json:
    input_file_path = path.sanitize(input_file_path)
    arg = shlex.split(
        'ffprobe -v quiet -print_format json -show_streams') + [input_file_path]
    out, _ = proc.run(arg)
    return json.loads(out)


def video_time(input_file_path: str) -> float:
    input_file_path = path.sanitize(input_file_path)
    arg = shlex.split(
        'ffprobe -v error -select_streams v:0 -show_entries format=duration -of default=noprint_wrappers=1:nokey=1') + [input_file_path]
    out, _ = proc.run(arg)
    return float(out)


def video_frame(input_file_path: str) -> int:
    input_file_path = path.sanitize(input_file_path)
    arg = shlex.split(
        'ffprobe -v error -select_streams v:0 -count_packets -show_entries stream=nb_read_packets -of csv=p=0') + [input_file_path]
    out, _ = proc.run(arg)
    return float(out)

File: src/ffprobe/ffprobe.go
package ffprobe

import (
	"log"
	"os/exec"
	"strconv"
	"strings"

	"github.com/sunrise2575/VP9-parallel/src/fsys"
	"github.com/tidwall/gjson"
)

func StreamInfoJSON(fp_in string) []gjson.Result {
	fp_in = fsys.Sanitize(fp_in)
	arg := strings.Fields("-v quiet -print_format json -show_streams")
	arg = append(arg, fp_in)

	result, e := exec.Command("ffprobe", arg...).Output()
	if e != nil {
		log.Panicf("[PANIC] Failed to execute ffprobe, error: %v", e)
	}

	return gjson.Get(string(result), "streams").Array()
}

func VideoTime(fp_in string) float64 {
	fp_in = fsys.Sanitize(fp_in)
	arg := strings.Fields("-v error -select_streams v:0 -show_entries format=duration -of default=noprint_wrappers=1:nokey=1")
	arg = append(arg, fp_in)

	bresult, e := exec.Command("ffprobe", arg...).Output()
	if e != nil {
		log.Panicf("[PANIC] Failed to execute ffprobe, error: %v", e)
	}

	result := strings.TrimRight(string(bresult), "\r\n")
	time, e := strconv.ParseFloat(result, 64)
	if e != nil {
		log.Panicf("[PANIC] strconv.ParseFloat(%v, 64) failed, error: %v", result, e)
	}

	return time
}

func VideoFrame(fp_in string) int {
	fp_in = fsys.Sanitize(fp_in)
	arg := strings.Fields("-v error -select_streams v:0 -count_packets -show_entries stream=nb_read_packets -of csv=p=0")
	arg = append(arg, fp_in)

	bresult, e := exec.Command("ffprobe", arg...).Output()
	if e != nil {
		log.Panicf("[PANIC] Failed to execute ffprobe, error: %v", e)
	}

	result := strings.TrimRight(string(bresult), "\r\n")
	frames, e := strconv.Atoi(result)
	if e != nil {
		log.Panicf("[PANIC] strconv.Atoi(%v) failed, error: %v", result, e)
	}

	return frames
}
